transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,698 @@
+        let udp_mux_read_fut = self
+            .udp_mux_read_fut
+            .get_or_insert(Box::pin(async move { udp_mux.read_from_conn().await }));
+        match udp_mux_read_fut.as_mut().poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(event) => {
+                self.udp_mux_read_fut = None;
+
+                match event {
+                    UDPMuxEvent::NewAddr(new_addr) => {
+                        let local_addr = socketaddr_to_multiaddr(&self.listen_addr);
+                        let send_back_addr = socketaddr_to_multiaddr(&new_addr.addr);
+                        let event = TransportEvent::Incoming {
+                            upgrade: Box::pin(upgrade::webrtc(
+                                self.udp_mux.clone(),
+                                self.config.clone(),
+                                new_addr.addr,
+                                new_addr.ufrag,
+                                self.id_keys.clone(),
+                            )) as BoxFuture<'static, _>,
+                            local_addr,
+                            send_back_addr,
+                            listener_id: self.listener_id,
+                        };
+                        Poll::Ready(Some(event))
+                    }
+                    UDPMuxEvent::Error(e) => {
+                        self.close(Err(Error::UDPMuxError(e)));
+                        return self.poll_next(cx);
+                    }
+                    _ => return self.poll_next(cx),
+                }
+            }
+        }
+    }
+}
+
+// TODO: remove
+fn hex_to_cow<'a>(s: &str) -> Cow<'a, [u8; 32]> {
+    let mut buf = [0; 32];
+    hex::decode_to_slice(s, &mut buf).unwrap();
+    Cow::Owned(buf)
+}
+
+/// Turns an IP address and port into the corresponding WebRTC multiaddr.
+pub(crate) fn socketaddr_to_multiaddr(socket_addr: &SocketAddr) -> Multiaddr {
+    // TODO: remove
+    let f = "ACD1E533EC271FCDE0275947F4D62A2B2331FF10C9DDE0298EB7B399B4BFF60B";
+    Multiaddr::empty()
+        .with(socket_addr.ip().into())
+        .with(Protocol::Udp(socket_addr.port()))
+        .with(Protocol::XWebRTC(hex_to_cow(&f)))
+}
+
+/// Extracts a SHA-256 fingerprint from the given address. Returns `None` if the address does not
+/// contain one.
+fn fingerprint_from_addr<'a>(addr: &'a Multiaddr) -> Option<Cow<'a, [u8; 32]>> {
+    let iter = addr.iter();
+    for proto in iter {
+        match proto {
+            // TODO: check hash is one of https://datatracker.ietf.org/doc/html/rfc8122#section-5
+            Protocol::XWebRTC(f) => return Some(f),
+            _ => continue,
+        }
+    }
+    None
+}
+
+/// Transforms a byte array fingerprint into a string.
+pub(crate) fn fingerprint_to_string<T>(f: T) -> String
+where
+    T: IntoIterator,
+    <T as IntoIterator>::Item: core::fmt::LowerHex,
+{
+    let values: Vec<String> = f.into_iter().map(|x| format! {"{:02x}", x}).collect();
+    values.join(":")
+}
```

Comment:
Have you considered introducing a `Fingerprint` newtype? That would allow us to remove the passing around of strings.

Reply:
I did, but waiting for https://github.com/multiformats/rust-multiaddr/pull/59 to be merged to see how much of the code will go away.

Reply:
That is orthogonal unless we define a newtype over there, is it not?

What I mean is, we extract the fingerprint here and pass it as a string to public APIs like `WebRTCConnection::accept` which relies on the string to have a specific format.